src/runtime_src/core/common/api/context_mgr.cpp

```diff
@@ -38,51 +28,57 @@ constexpr auto virtual_cu_idx = std::numeric_limits<unsigned int>::max();
 // the former has closed its context.
 class device_context_mgr
 {
-  std::mutex m_mutex;
-  std::condition_variable m_cv;
-  xrt_core::device* m_device;
-
-  // CU contexts are managed per domain where indicies are
-  // in the range [0..max_cus[.  This allows for a compact
-  // bitset representation.  CU indices are unique within
-  // a domain and global for all hardware contexts.
-  using domain_type = cuidx_type::domain_type;
-  std::map<domain_type, std::bitset<max_cus>> m_d2ctx; // domain -> cxt
-
-  // Get the domain context bitset that contains this ipidx
-  // Return by address since used conditionally in get_idx_ctx
-  std::bitset<max_cus>*
-  get_ctx(cuidx_type ipidx)
+  // CU indeces are managed per hwctx
+  // This struct manages CUs are that are opened by the
+  // context mananger.  It supports mapping
+  // - {ctx, nm} -> ip_info   // for opening
+  // - {ctx, idx} -> ip_info  // for closing
+  // where the ip_info data is shared by both maps
+  struct ctx
   {
-    return &(m_d2ctx[ipidx.domain]);
-  }
+    struct ip {
+      std::string ipname;
+      cuidx_type ipidx;
+
+      ip(const std::string& nm, cuidx_type idx)
+        : ipname(nm), ipidx(idx)
+      {}
+    };
+
+    std::map<std::string, std::shared_ptr<ip>> m_nm2ip;
+    std::map<decltype(cuidx_type::index), std::shared_ptr<ip>> m_idx2ip;
+
+    const ip*
+    get(const std::string& ipname)
+    {
+      return m_nm2ip[ipname].get();
```

Comment:
what if ipname doesn't exists in m_nm2ip map. please add code for ipname  check.